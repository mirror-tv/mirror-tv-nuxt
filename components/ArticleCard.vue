<template>
  <div>
    <ArticleCardColumnInMobile
      v-if="mobileLayoutDirection === 'column'"
      v-bind="$props"
    />
    <ArticleCardRowInMobile
      v-else-if="mobileLayoutDirection === 'row'"
      v-bind="$props"
    />
  </div>
</template>

<script>
import ArticleCardColumnInMobile from './ArticleCardColumnInMobile'
import ArticleCardRowInMobile from './ArticleCardRowInMobile'

export default {
  components: {
    ArticleCardColumnInMobile,
    ArticleCardRowInMobile,
  },
  props: {
    href: {
      type: String,
      default: '',
    },
    labelTitle: {
      type: String,
      default: '',
    },
    articleTitleStyle: {
      type: String,
      default: 'normal',
    },
    articleTitleHighlightText: {
      type: String,
      default: '',
    },
    articleImgURL: {
      type: String,
      default: require('~/assets/img/default/image-default.jpg'),
    },
    articleTitle: {
      type: String,
      default: '',
    },
    articleDescription: {
      type: String,
      default: '',
    },
    articleDate: {
      type: Date,
      default: () => new Date(),
    },
    articleStyle: {
      type: String,
      default: null,
    },
    isMicroAd: {
      type: Boolean,
      default: false,
    },
    microAdId: {
      type: String,
      default: '',
    },

    // Differentiate two usages in / and /category/:name pages
    mobileLayoutDirection: {
      type: String,
      default: 'row',
      validator(value) {
        return ['row', 'column'].includes(value)
      },
    },
  },
}
</script>

<style lang="scss" scoped>
div {
  height: 100%;
}
</style>
