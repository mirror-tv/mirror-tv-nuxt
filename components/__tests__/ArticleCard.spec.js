import { shallowMount } from '@vue/test-utils'
import ArticleCard from '../ArticleCard'

describe('props', function() {
  test('Should hide .label-title-wrapper if we did not provide "labelTitle" props', function() {
    const wrapper = shallowMount(ArticleCard)
    const labelTitleWrapper = wrapper.find('.label-title-wrapper')
    expect(labelTitleWrapper.exists()).toBe(false)
  })

  test('Should render text by "labelTitle" props', function() {
    const labelTitleMock = 'labelTitleMock'
    const wrapper = shallowMount(ArticleCard, {
      propsData: {
        labelTitle: labelTitleMock
      }
    })
    const barTitle = wrapper.find('.label-title')
    expect(barTitle.text()).toBe(labelTitleMock)
  })

  test('Should render image src by "articleImgURL" props', async function() {
    expect.assertions(1)
    const articleImgURLMock = 'articleImgURLMock'
    const wrapper = await shallowMount(ArticleCard, {
      propsData: {
        articleImgURL: articleImgURLMock
      }
    })
    const articleImg = wrapper.find('.article-img')
    expect(articleImg.attributes()['data-src']).toBe(articleImgURLMock)
  })

  test('Should render text by "articleTitle" props', function() {
    const articleTitleMock = 'articleTitleMock'
    const wrapper = shallowMount(ArticleCard, {
      propsData: {
        articleTitle: articleTitleMock
      }
    })
    const articleTitle = wrapper.find('.article-title')
    expect(articleTitle.text()).toBe(articleTitleMock)
  })

  test('Should render text by "articleDate" props', function() {
    const articleDateMockString = '2020/01/02 03:04'
    const articleDateMock = new Date(2020, 0, 2, 3, 4)
    const wrapper = shallowMount(ArticleCard, {
      propsData: {
        articleDate: articleDateMock
      }
    })
    const articleDate = wrapper.find('.article-date')
    expect(articleDate.text()).toBe(articleDateMockString)
  })

  test('Should render href by "href" props', function() {
    const hrefMock = 'https://www.google.com'
    const wrapper = shallowMount(ArticleCard, {
      propsData: {
        href: hrefMock
      }
    })
    expect(wrapper.attributes().href).toBe(hrefMock)
  })

  test('Should have class "row-mobile" on .bottom-wrapper if we did not provide "mobileLayoutDirection" props', function() {
    const wrapper = shallowMount(ArticleCard)
    const bottomWrapper = wrapper.find('.bottom-wrapper')
    expect(bottomWrapper.classes()).toContain('row-mobile')
  })

  test('Should have class "shrink" on .article-img if we did not provide "mobileLayoutDirection" props', function() {
    const wrapper = shallowMount(ArticleCard)
    const articleImage = wrapper.find('.article-img')
    expect(articleImage.classes()).toContain('shrink')
  })

  test('Should have modifier "margin-left" on .bottom-wrapper__info-wrapper if we did not provide "mobileLayoutDirection" props', function() {
    const wrapper = shallowMount(ArticleCard)
    const infoWrapper = wrapper.find('.bottom-wrapper__info-wrapper')
    expect(infoWrapper.classes()).toContain(
      'bottom-wrapper__info-wrapper--margin-left'
    )
  })

  test('Should have class "column-mobile" on .bottom-wrapper if we provide "column" to "mobileLayoutDirection" props', function() {
    const wrapper = shallowMount(ArticleCard, {
      propsData: {
        mobileLayoutDirection: 'column'
      }
    })
    const bottomWrapper = wrapper.find('.bottom-wrapper')
    expect(bottomWrapper.classes()).toContain('column-mobile')
  })

  test('Should have class "stretch" on .article-img if we provide "column" to "mobileLayoutDirection" props', function() {
    const wrapper = shallowMount(ArticleCard, {
      propsData: {
        mobileLayoutDirection: 'column'
      }
    })
    const articleImage = wrapper.find('.article-img')
    expect(articleImage.classes()).toContain('stretch')
  })

  test('Should have modifier "margin-right" on .bottom-wrapper__info-wrapper if we provide "column" to "mobileLayoutDirection" props', function() {
    const wrapper = shallowMount(ArticleCard, {
      propsData: {
        mobileLayoutDirection: 'column'
      }
    })
    const infoWrapper = wrapper.find('.bottom-wrapper__info-wrapper')
    expect(infoWrapper.classes()).toContain(
      'bottom-wrapper__info-wrapper--margin-right'
    )
  })
})

describe('snapshot tests', function() {
  test('Usage with provided labelTitle', async function() {
    const wrapper = await shallowMount(ArticleCard, {
      propsData: {
        href: 'https://www.mirrormedia.mg/story/20200715edi026/',
        labelTitle: '生活',
        articleImgURL:
          'https://www.mirrormedia.com.tw/assets/images/20200715135810-51b41bed253ae37fe37c5f9972844e09-mobile.jpg',
        articleTitle: '振興券上路最大咖好禮　黃偉哲：來台南旅遊消費抽房子',
        articleDate: new Date(2020, 0, 2, 3, 4, 5)
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  test('Usage with not provided labelTitle', async function() {
    const wrapper = await shallowMount(ArticleCard, {
      propsData: {
        href: 'https://www.mirrormedia.mg/story/20200715edi026/',
        // labelTitle: '生活',
        articleImgURL:
          'https://www.mirrormedia.com.tw/assets/images/20200715135810-51b41bed253ae37fe37c5f9972844e09-mobile.jpg',
        articleTitle: '振興券上路最大咖好禮　黃偉哲：來台南旅遊消費抽房子',
        articleDate: new Date(2020, 0, 2, 3, 4, 5)
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  test('Usage with provided mobileLayoutDirection', async function() {
    const wrapper = await shallowMount(ArticleCard, {
      propsData: {
        href: 'https://www.mirrormedia.mg/story/20200715edi026/',
        labelTitle: '生活',
        articleImgURL:
          'https://www.mirrormedia.com.tw/assets/images/20200715135810-51b41bed253ae37fe37c5f9972844e09-mobile.jpg',
        articleTitle: '振興券上路最大咖好禮　黃偉哲：來台南旅遊消費抽房子',
        articleDate: new Date(2020, 0, 2, 3, 4, 5),
        mobileLayoutDirection: 'column'
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
