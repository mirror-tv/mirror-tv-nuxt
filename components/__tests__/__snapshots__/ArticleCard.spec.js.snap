// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests Usage with not provided labelTitle 1`] = `
<div>
  <a
    articledescription=""
    articletitlehighlighttext=""
    articletitlestyle="normal"
    class="article-card"
    href="https://www.mirrormedia.mg/story/20200715edi026/"
    mobilelayoutdirection="row"
    rel="noreferrer noopener"
    target="_blank"
  >
    <span
      class="top-wrapper"
    >
      <!---->
    </span>
     
    <span
      class="article-card__bottom-wrapper bottom-wrapper row-mobile"
    >
      <span
        class="article-img-wrapper shrink"
      >
        <img
          alt="article-img"
          class="article-img"
          data-src="https://www.mirrormedia.com.tw/assets/images/20200715135810-51b41bed253ae37fe37c5f9972844e09-mobile.jpg"
          lazy="loading"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        />
      </span>
       
      <span
        class="bottom-wrapper__info-wrapper bottom-wrapper__info-wrapper--margin-left info-wrapper"
      >
        <span
          class="article-title"
        >
          振興券上路最大咖好禮　黃偉哲：來台南旅遊消費抽房子
        </span>
         
        <span
          class="info-wrapper__article-date article-date"
        >
          2020/01/02 03:04
        </span>
      </span>
    </span>
  </a>
</div>
`;

exports[`snapshot tests Usage with provided labelTitle 1`] = `
<div>
  <a
    articledescription=""
    articletitlehighlighttext=""
    articletitlestyle="normal"
    class="article-card"
    href="https://www.mirrormedia.mg/story/20200715edi026/"
    mobilelayoutdirection="row"
    rel="noreferrer noopener"
    target="_blank"
  >
    <span
      class="top-wrapper"
    >
      <span
        class="top-wrapper__label-title-wrapper label-title-wrapper"
      >
        <span
          class="label-title"
        >
          生活
        </span>
      </span>
    </span>
     
    <span
      class="article-card__bottom-wrapper bottom-wrapper row-mobile"
    >
      <span
        class="article-img-wrapper shrink"
      >
        <img
          alt="article-img"
          class="article-img"
          data-src="https://www.mirrormedia.com.tw/assets/images/20200715135810-51b41bed253ae37fe37c5f9972844e09-mobile.jpg"
          lazy="loading"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        />
      </span>
       
      <span
        class="bottom-wrapper__info-wrapper bottom-wrapper__info-wrapper--margin-left info-wrapper"
      >
        <span
          class="article-title"
        >
          振興券上路最大咖好禮　黃偉哲：來台南旅遊消費抽房子
        </span>
         
        <span
          class="info-wrapper__article-date article-date"
        >
          2020/01/02 03:04
        </span>
      </span>
    </span>
  </a>
</div>
`;

exports[`snapshot tests Usage with provided mobileLayoutDirection 1`] = `
<div>
  <a
    class="article-card"
    href="https://www.mirrormedia.mg/story/20200715edi026/"
    labeltitle="生活"
    mobilelayoutdirection="column"
    rel="noreferrer noopener"
    target="_blank"
  >
    <span
      class="article-card__bottom-wrapper bottom-wrapper column-mobile"
    >
      <span
        class="article-img-wrapper stretch"
      >
        <img
          alt="article-img"
          class="article-img"
          data-src="https://www.mirrormedia.com.tw/assets/images/20200715135810-51b41bed253ae37fe37c5f9972844e09-mobile.jpg"
          lazy="loading"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        />
      </span>
       
      <span
        class="bottom-wrapper__info-wrapper bottom-wrapper__info-wrapper--margin-right info-wrapper"
      >
        <span
          class="article-title"
          style="font-weight: normal;"
        >
          振興券上路最大咖好禮　黃偉哲：來台南旅遊消費抽房子
        </span>
         
        <span
          class="article-description"
        />
         
        <span
          class="info-wrapper__article-date article-date"
        >
          2020/01/02 03:04
        </span>
      </span>
    </span>
  </a>
</div>
`;
